File: PDDL.py
# -*- coding: utf-8 -*-
"""
@author: Blopa
"""

### ERROR
class PDDLSyntaxError(Exception):
	pass
class MissingRequirementError(Exception):
	pass
class UnclosedExpressionError(Exception):
	pass
class UnknownDomainError(Exception):
	pass
class UnknownTypeError(Exception):
	pass

def parseSExpression(s):
	tokens = s.replace('(',' ( ').replace(')',' ) ').replace(',',' ').replace('"',' ').replace('\n',' ').replace('\t','').split(' ')
	stack = []
	for token in tokens:
		if not token or token==' ':
			continue
		elif token==')':
			expr = []
			while stack and stack[-1]!='(':
				expr.insert(0,stack.pop())
			if stack:
				stack.pop()
				stack.append(expr)
			else:
				raise UnclosedExpressionError
		else:
			stack.append(token)
	return stack.pop()

support = [':adl',':strips',':typing',':disjunctive-preconditions',':equality',':existential-preconditions',':universal-preconditions',':quantified-preconditions',':conditional-effects']
domains = {}

def parse_typed_list(typed):
	objects = {}
	batch = []
	typing = False
	for obj in typed:
		if len(obj)==0:
			continue
		elif obj=='-':
			typing = True
		elif typing:
			typing = False
			while batch:
				objects[batch.pop()] = obj
		else:
			batch.append(obj)
	while batch:
		objects[batch.pop()] = ''
	return objects

def parse_domain(filename):
	with open(filename,'r') as file:
		try:
			pddl = ' '.join([line.replace('\n',' ') for line in file.readlines() if not ';;' in line])	
			domain_tree = parseSExpression(pddl)
			if domain_tree[0]!='define' or domain_tree[1][0]!='domain':
				raise PDDLSyntaxError
			domain_name = domain_tree[1][1]
			reqs = ''
			known_types = {'':''}
			known_constants = {}
			known_predicates = {}
			known_actions = {}
			for section in domain_tree:
				if section[0]==':requirements':
					reqs = section[1:]
					for req in reqs:
						if req not in support:
							raise MissingRequirementError
				elif section[0]==':types':
					known_types = parse_typed_list(section[1:])
					known_types[''] = ''
					types = list(known_types.keys())
					for subtype in types:
						if known_types[subtype] not in known_types:
							known_types[known_types[subtype]] = ''
				elif section[0]==':constants':
					known_constants = parse_typed_list(section[1:])
				elif section[0]==':predicates':
					for pred in section[1:]:
						known_predicates[pred[0]] = parse_typed_list(pred[1:])
				elif section[0]==':action':
					action_name = section[1]
					parameters = {}
					precondition = None
					effect = None
					for s in range(2,len(section)):
						if section[s]==':parameters':
							parameters = parse_typed_list(section[s+1])
						elif section[s]==':precondition':
							precondition = section[s+1]
						elif section[s]==':effect':
							effect = section[s+1]
					known_actions[action_name] = {
							'parameters': parameters,
							'precondition': precondition,
							'effect': effect
					}
			domains[domain_name] = {
					'name': domain_name,
					'requirements': reqs,
					'types': known_types,
					'constants': known_constants,
					'predicates': known_predicates,
					'actions': known_actions
			}
			return domains[domain_name]
		except:
			raise PDDLSyntaxError	
		
def parse_problem(filename):
	with open(filename,'r') as file:
		try:
			pddl = ' '.join([line.replace('\n',' ') for line in file.readlines() if not ';;' in line])	
			problem_tree = parseSExpression(pddl)
			if problem_tree[0]!='define' or problem_tree[1][0]!='problem':
				raise PDDLSyntaxError
			problem_name = problem_tree[1][1]
			domain_name = ''
			known_objects = []
			known_atoms = []
			known_goal = None
			for section in problem_tree:
				if section[0]==':domain':
					domain_name = section[1]
				elif section[0]==':objects':
					objects = []
					typing = False
					for obj in section[1:]:
						if len(obj)==0:
							continue
						elif obj=='-':
							typing = True
						elif typing:
							typing = False
							while objects:
								known_objects.append([objects.pop(),obj])
						else:
							objects.append(obj)
					while objects:
						known_objects.append([objects.pop(),''])
				elif section[0]==':init':
					for a in section[1:]:
						known_atoms.append(a)
				elif section[0]==':goal':
					known_goal = section[1]
			if domain_name not in domains.keys():
				raise UnknownDomainError
			domain = domains[domain_name]
			known_objects.extend([[key,domain['constants'][key]] for key in domain['constants']])
			known_sets = {key: [] for key in domain['types'].keys()}
			for obj in known_objects:
				if obj[1] not in known_sets.keys():
					raise UnknownTypeError
				known_sets[obj[1]].append(obj[0])
			for i in range(len(domain['types'])):
				for typ in domain['types']:
					for obj in known_sets[typ]:
						if obj not in known_sets[domain['types'][typ]]:
							known_sets[domain['types'][typ]].append(obj)
			problem = {
				'name': problem_name,
				'domain': domain,
				'atoms': known_atoms,
				'sets': known_sets,
				'goal': known_goal
			}
			return problem
		except:
			raise PDDLSyntaxError
			
			
if __name__ == "__main__":
	d = parse_domain('classical-domains/classical/miconic/domain.pddl')
	p = parse_problem('classical-domains/classical/miconic/s1-0.pddl')
	dx = parse_domain('classical-domains/classical/airport/p06-domain.pddl')
	px = parse_problem('classical-domains/classical/airport/p06-airport2-p2.pddl')

File: PDDL.cpp
#ifndef PDDL_CPP
#define PDDL_CPP
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

namespace PDDL{
	
	class Domain;
	class Problem;
	
	std::string loadFile(const std::string &filename);
	std::vector<std::string> splitSections(const std::string &s);
	std::string flattenString(std::string s);
	Domain* parsePDDLDomain(const std::string &filename);
	Problem* parsePDDLProblem(const std::string &filename);
	
	const char* supported[] = {":strips",":typing",":disjunctive-preconditions",
								":equality",":existential-preconditions",
								":universal-preconditions",":quantified-preconditions",
								":conditional-effects",":adl", 0};
	
	class Domain{
		protected:
			static std::map<std::string,Domain*> domains;
		public:
			class Action{
				public:
					std::string name;
					std::string precondition;
					std::string effect;
					std::vector<std::pair<std::string,std::string>> parameters;
					friend std::ostream& operator<<(std::ostream &out, const Action &a);
			};
			std::string domain;
			std::vector<std::string> requirements;
			std::map<std::string,std::set<std::string>> itypes;
			std::map<std::string,std::set<std::string>> types;
			std::vector<Action> actions;
			friend std::ostream& operator<<(std::ostream &out, Domain &d);
			friend Domain* parsePDDLDomain(const std::string &filename);
			friend Problem* parsePDDLProblem(const std::string &filename);
	};
	
	class Problem{
		protected:
			static std::map<std::pair<std::string,std::string>,Problem*> problems;
		public:
			Domain* domain;
			std::string problem;
			std::string goal;
			std::set<std::string> init;
			std::map<std::string,std::set<std::string>> sets;
			friend std::ostream& operator<<(std::ostream &out, Problem &p);
			friend Problem* parsePDDLProblem(const std::string &filename);
	};
	
	// Domain class
	std::map<std::string,Domain*> Domain::domains;
	
	std::ostream& operator<<(std::ostream &out, const Domain::Action &a){
		int i=0;
		out << "\t(:action " << a.name << std::endl << "\t\t:parameters (";
		for(const std::pair<std::string,std::string> &param : a.parameters){ out << (i++?" ":"") << param.first << " - " << param.second; }
		return out << ")" << std::endl << "\t\t:precondition (" << a.precondition << ")" << std::endl << "\t\t:effect (" << a.effect << ")" << std::endl << "\t)";
	}
	
	std::ostream& operator<<(std::ostream &out, Domain &d){
		out << "(define " << std::endl;
		out << "\t(domain " << d.domain << ")" << std::endl;
		if(!d.requirements.empty()){
			out << "\t(:requirements";
			for(const std::string &s : d.requirements){ out << " " << s; }
			out << ")" << std::endl;
		}
		if(!d.types.empty()){
			out << "\t(:types" << std::endl;
			for(const std::pair<const std::string, const std::set<std::string>> &type : d.types){
				out <<"\t\t";
				for(const std::string &typ : type.second){ out << typ << " "; }
				out << "- " << type.first << std::endl;
			}
			out << "\t)" << std::endl;
		}
		for(Domain::Action &a : d.actions){ out << a << std::endl; }
		return out << ")";
	}
	
	// Problem class
	std::map<std::pair<std::string,std::string>,Problem*> Problem::problems;
	
	std::ostream& operator<<(std::ostream &out, Problem &p){
		out << "(define " << std::endl;
		out << "\t(problem " << p.problem << ")" << std::endl << "\t(:domain " << p.domain->domain << ")" << std::endl << "\t(:objects" << std::endl;
		for(const std::pair<std::string,std::set<std::string>> &sets : p.sets){
			out << "\t\t";
			for(const std::string &obj : sets.second){
				out << obj << " ";
			}
			out << "- " << sets.first << std::endl;
		}
		out << "\t)" << std::endl << "\t(:init" << std::endl;
		for(const std::string &atom : p.init){
			out << "\t\t(" << atom << ")" << std::endl;
		}
		return out << "\t)" << std::endl << "\t(:goal (" << p.goal << "))" << std::endl << ")";
	}
	
	// global functions
	std::string loadFile(const std::string &filename){
		std::ifstream file(filename);
		std::stringstream buffer;
		buffer << file.rdbuf();
		return buffer.str();
	}
	
	std::vector<std::string> splitSections(const std::string &s){
		std::vector<std::string> sections;
		size_t anchor;
		for(size_t reader = 0;reader<s.size();){
			anchor = reader;
			if(s.at(reader)==' ' || s.at(reader)=='\n' || s.at(reader)=='\r'){
				reader++;
				continue;
			}else if(s.at(reader)=='('){
				anchor++;
				for(int pars=1;pars;){
					reader++;
					if(s[reader]=='('){ ++pars; }
					if(s[reader]==')'){ --pars; }
				}
			}else if(s.at(reader)==';' && s.at(reader+1)==';'){
				while(reader<s.size() && s.at(reader)!='\n' && s.at(reader)!='\r'){ reader++; }
			}else{
				while(reader<s.size() && s.at(reader)!=' ' && s.at(reader)!='\n' && s.at(reader)!='\r'){ reader++; }
			}
			sections.push_back(s.substr(anchor,reader-anchor));
			reader++;
		}
		return sections;
	}
	
	std::string flattenString(std::string s){
		std::string result;
		bool flag = false;
		for(char c : s){
			if(c!='\n' && c!='\r' && c!=' '){
				if(flag){ result += ' '; }
				result += c;
				flag = false;
			}else if(!flag){
				flag = true;
			}
		}
		return result;
	}
	
	// TODO: Make safe -> adds exceptions/errors
	Domain* parsePDDLDomain(const std::string &filename){
		Domain* domain = 0;
		std::string domainStr = loadFile(filename);
		size_t lPar = domainStr.find_first_of('(');
		size_t anchor = domainStr.find_last_of(')');
		domainStr = domainStr.substr(lPar+1, anchor-(lPar+1));
		std::vector<std::string> sections = splitSections(domainStr);
		for(std::string &section : sections){
			if(section=="define" || (section.at(0)==';' && section.at(1)==';')){ continue; }
			std::vector<std::string> subsections = splitSections(section);
			if(subsections.empty()){ continue; }
			if(subsections[0]=="domain"){
				Domain** domainPtr = &Domain::domains[subsections[1]];
				if(!*domainPtr){
					*domainPtr = new Domain();
					domain = *domainPtr;
				}else{
					domain = *domainPtr;
					break;
				}
				domain->domain = subsections[1];
			}
			if(!domain){ continue; }
			if(subsections[0]==":extends"){
				Domain* domainPtr = Domain::domains.at(subsections[1]);
				if(domainPtr){
					domain->requirements = domainPtr->requirements;
					domain->types = domainPtr->types;
					domain->itypes = domainPtr->itypes;
					domain->actions = domainPtr->actions;
					// domain->constants = domainPtr->constants;
					// domain->predicates = domainPtr->predicates;
				}else{
					// TODO: Throw exception
				}
			}
			if(subsections[0]==":requirements"){
				domain->requirements = std::vector<std::string>(++subsections.begin(),subsections.end());
				// TODO: Verify requirements, throw exception
			}
			if(subsections[0]==":types"){
				std::vector<std::string> types;
				bool parent = false;
				for(int i=1;i<subsections.size();i++){
					if(subsections[i]=="-"){
						parent = true;
					}else if(parent){
						for(const std::string &typ : types){
							domain->types[subsections[i]].insert(typ);
							domain->itypes[typ].insert(subsections[i]);
						}
						types.clear();
						parent = false;
					}else{
						types.push_back(subsections[i]);
					}
				}
				for(const std::string &typ : types){ domain->itypes[typ]; }
			}
			if(subsections[0]==":constants"){
				// TODO: Implement, constant objects present in all problems of the domain
			}
			if(subsections[0]==":action"){
				Domain::Action newAction;
				newAction.name = subsections[1];
				for(int i=2;i<subsections.size()-1;i++){
					if(subsections[i]==":parameters"){
						std::vector<std::string> params = splitSections(subsections[++i]);
						std::vector<std::string> queue;
						bool type = false;
						for(int j=0;j<params.size();j++){
							if(params[j]=="-"){
								type=true;
							}else if(type){
								// TODO: Add support for *either*
								type = false;
								for(const std::string &elem : queue){ newAction.parameters.push_back({elem,params[j]}); }
								queue.clear();
							}else{
								queue.push_back(params[j]);
							}
						}
						for(const std::string &elem : queue){ newAction.parameters.push_back({elem,""}); }
					}else if(subsections[i]==":precondition"){
						newAction.precondition = flattenString(subsections[++i]);
					}else if(subsections[i]==":effect"){
						newAction.effect = flattenString(subsections[++i]);
					}
				}
				domain->actions.push_back(newAction);
			}
			if(subsections[0]==":domain-variables" || subsections[0]==":timeless" || subsections[0]==":axiom" || subsections[0]==":safety"){
				// TODO: Throw unsupported
			}
			if(subsections[0]==":predicates"){
				// TODO: Support predicates
			}
		}
		return domain;
	}
	
	Problem* parsePDDLProblem(const std::string &filename){
		Problem* problem = 0;
		std::string problemStr = loadFile(filename);
		size_t lPar = problemStr.find_first_of('(');
		size_t anchor = problemStr.find_last_of(')');
		problemStr = problemStr.substr(lPar+1, anchor-(lPar+1));
		std::vector<std::string> sections = splitSections(problemStr);
		std::string tmpStr;
		for(std::string &section : sections){
			if(section=="define" || (section.at(0)==';' && section.at(1)==';')){ continue; }
			std::vector<std::string> subsections = splitSections(section);
			if(subsections.empty()){ continue; }
			if(subsections[0]=="problem" || subsections[0]==":domain"){
				if(tmpStr.empty()){
					tmpStr = subsections[1];
				}else{
					std::pair<std::string,std::string> key = (subsections[0].at(0)==':')?std::pair<std::string,std::string>{subsections[1],tmpStr}:std::pair<std::string,std::string>{tmpStr,subsections[1]};
					Problem** problemPtr = &Problem::problems[key];
					if(!*problemPtr){
						*problemPtr = new Problem();
						(*problemPtr)->domain = Domain::domains.at((subsections[0].at(0)==':')?subsections[1]:tmpStr);
						problem = *problemPtr;
					}else{
						problem = *problemPtr;
						break;
					}
				}
			}
			if(!problem){ continue; }
			if(subsections[0]==":objects"){
				std::vector<std::string> queue;
				bool type = false;
				for(int i=1;i<subsections.size();i++){
					if(subsections[i]=="-"){
						type=true;
					}else if(type){
						type = false;
						for(const std::string &elem : queue){
							problem->sets[subsections[i]].insert(elem);
							problem->sets[""].insert(elem);
						}
						queue.clear();
					}else{
						queue.push_back(subsections[i]);
					}
				}
				for(const std::string &elem : queue){ problem->sets[""].insert(elem); }
				for(int i=0;i<problem->domain->types.size();i++){
					for(const std::pair<std::string,std::set<std::string>> &set : problem->domain->types){
						for(const std::string &type : set.second){
							for(const std::string &elem : problem->sets.at(type)){
								problem->sets[set.first].insert(elem);
							}
						}
					}
				}
			}
			if(subsections[0]==":init"){
				for(int i=1;i<subsections.size();i++){
					problem->init.insert(subsections.at(i));
				}
			}
			if(subsections[0]==":goal"){
				problem->goal = flattenString(subsections[1]);
			}
		}
		return problem;
	}
};

#endif
